net.solarnetwork.node.hw.energymeter.mock/src/net/solarnetwork/node/datum/energymeter/mock/MockEnergyMeterDatumSource.java

```diff
@@ -0,0 +1,204 @@
+package net.solarnetwork.node.datum.energymeter.mock;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import org.springframework.context.MessageSource;
+
+import net.solarnetwork.node.DatumDataSource;
+import net.solarnetwork.node.domain.GeneralNodeACEnergyDatum;
+import net.solarnetwork.node.settings.SettingSpecifier;
+import net.solarnetwork.node.settings.SettingSpecifierProvider;
+import net.solarnetwork.node.settings.support.BasicTextFieldSettingSpecifier;
+import net.solarnetwork.node.settings.support.BasicToggleSettingSpecifier;
+
+/**
+ * Mock plugin to be the source of values for a GeneralNodeACEnergyDatum, the
+ * datum will contain random values based around a settable center deviating by
+ * a settable deviation.
+ * 
+ * <p>
+ * This class implements {@link SettingSpecifierProvider} and
+ * {@link DatumDataSource}
+ * </p>
+ * 
+ * @author robert
+ * @version 1.0
+ */
+
+public class MockEnergyMeterDatumSource implements DatumDataSource<GeneralNodeACEnergyDatum>, SettingSpecifierProvider {
+
+	@Override
+	public String getUID() {
+		return null;
+	}
+
+	@Override
+	public String getGroupUID() {
+		return null;
+	}
+
+	@Override
+	public Class<? extends GeneralNodeACEnergyDatum> getDatumType() {
+		return GeneralNodeACEnergyDatum.class;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see net.solarnetwork.node.DatumDataSource#readCurrentDatum()
+	 * 
+	 * Returns a {@link GeneralNodeACEnergyDatum} filled with mock data. The
+	 * method uses reflection to find setters in {@link
+	 * GeneralNodeACEnergyDatum} and fill them with a number equal to the
+	 * baseReading +- a random number that goes upto deviation.
+	 * 
+	 * @return A {@link GeneralNodeACEnergyDatum}
+	 * 
+	 */
+	@Override
+	public GeneralNodeACEnergyDatum readCurrentDatum() {
+
+		GeneralNodeACEnergyDatum datum = new GeneralNodeACEnergyDatum();
+		datum.setCreated(new Date());
+		datum.setSourceId(sourceId);
+
+		Class<GeneralNodeACEnergyDatum> gacdatum = GeneralNodeACEnergyDatum.class;
+
+		Method[] gacdMethods = gacdatum.getMethods();
```

Comment:
Using reflection and random values for all properties here seems odd to me. We know exactly the type of data we're mocking here, this class should simply use the methods on `GeneralNodeACEnergyDatum` to generate fake, but reasonable, values for the properties it wants to mimic... I'd like to think of this class as a power meter _simulator_ more than a random number generator.

The OCCP code has [an example of a mock power meter](https://github.com/SolarNetwork/solarnetwork-node/blob/45338407ea7d81ea2a7afdcfa72eefcc10485314/net.solarnetwork.node.ocpp.mock/src/net/solarnetwork/node/ocpp/mock/MockMeterDataSource.java#L60) that might be a good reference. The point is to try to simulate the characteristics of a real power meter, so in that class you'll see that it randomizes the `watts` of each sample, but calculates actual `wattHourReading` values from those randomized samples across time. You'll also see how that class can simulate "charging" vs "idle" modes, by ramping up the randomized `watt` values when "charging" is enabled. That is similar to what you're doing with the `deviate` setting in this class.

So `watts` and `wattHourReading`  the most important properties to simulate in a power meter, but if others are needed they should try to mimic the actual characteristics of power meters.

Reply:
The reason it was all random values before was I just needed values to plot on my graph. I have tried to follow your advice and have changed it to be a simulator of a resister and inductor in series.